include/ghex/transport_layer/ucx/context.hpp

```diff
@@ -175,69 +182,68 @@ namespace gridtools {
-                        reinterpret_cast<unsigned char*>(worker_address),
-                        reinterpret_cast<unsigned char*>(worker_address) + address_length};
-                    ucp_worker_release_address(m_worker.get(), worker_address);
-                    m_worker.m_moved = false;
-                }
-                
+                    : m_context{c}
+                    , m_parallel_context{pc}
+                    , m_token_ptr{t}
+                    , m_rank(c->rank())
+                    , m_size(c->size())
+                    {
+                        ucp_worker_params_t params;
+                        params.field_mask  = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+                        params.thread_mode = mode;
+                        GHEX_CHECK_UCX_RESULT(
+                            ucp_worker_create (c->get(), &params, &m_worker.get())
+                        );
+                        ucp_address_t* worker_address;
+                        std::size_t address_length;
+                        GHEX_CHECK_UCX_RESULT(
+                            ucp_worker_get_address(m_worker.get(), &worker_address, &address_length)
+                        );
+                        m_address = address_t{
+                            reinterpret_cast<unsigned char*>(worker_address),
+                            reinterpret_cast<unsigned char*>(worker_address) + address_length};
+                        ucp_worker_release_address(m_worker.get(), worker_address);
+                        m_worker.m_moved = false;
+                    }
+
                 template<typename ThreadPrimitives>
                 const endpoint_t& worker_t<ThreadPrimitives>::connect(rank_type rank)
                 {
                     auto it = m_endpoint_cache.find(rank);
                     if (it != m_endpoint_cache.end())
                         return it->second;
-                    if (auto addr_ptr = m_context->m_db.find(rank))
-                    {
-                        auto p = m_endpoint_cache.insert(std::make_pair(rank, endpoint_t{rank, m_worker.get(), *addr_ptr}));
-                        return p.first->second;
-                    }
-                    else
-                        throw std::runtime_error("could not connect to endpoint");
+                    auto addr = m_context->m_db.find(rank);
+                    auto p = m_endpoint_cache.insert(std::make_pair(rank, endpoint_t{rank, m_worker.get(), addr}));
+                    return p.first->second;
                 }
 
             } // namespace ucx
-            
+
             template<class ThreadPrimitives>
             struct context_factory<ucx_tag, ThreadPrimitives>
             {
-                static std::unique_ptr<context<ucx_tag, ThreadPrimitives>> create(int num_threads, MPI_Comm mpi_comm)
+                static std::unique_ptr<context<ucx_tag, ThreadPrimitives>> create(int num_threads, MPI_Comm mpi_comm = MPI_COMM_NULL)
                 {
```

Comment:
Maybe we should have 2 create functions, second one:
```
template<class DB>
static auto create(int num_threads, DB&& db)
```

Reply:
I guess we could even only have the second version, and always require that the user specifies the db himself. We should probably also check, whether `m_world.m_comm` is non-null in the `context::barrier`. otherwise the barrier silently doesn't work.